keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,812 @@
+* without the need of semantical understanding of **all** OpenAPI v3 constructs.
+
+Here is an example of what we want to avoid:
+
+```yaml
+properties:
+  foo:
+    type: string
+  bar:
+    minimum: 42
+anyOf:
+- properties:
+    foo:
+      pattern: abc
+    bar:
+      type: string
+- properties:
+    bar:
+      type: integer
+      minimum: 1
+      maximum: 0
+```
+In this example, `.foo` is specified as string even without looking at the `anyOf` construct, but `.bar`'s type is unclear without deeper semantic knowledge of `anyOf` and of minimum and maximum.
+
+We want that specified types and object field names are mechanically extractable with an understanding of `properties`, `items`, `additionaProperties` and `type` only. Other constructs like logical junctors should be usable to express complex validations, but should not be essential for the specification of the type structure of JSON values.
+
+We reach this goal in defining _structural schema_ by following this guiding principle:
+
+**Structural Completeness**: all types and the possible object fields within a JSON object must be apparent from a structural schema without understanding the logical junctors `anyOf`, `allOf`, `oneOf` and `not`.
+
+### Formal Definition
+
+We define a _structural schema_ by choosing another data structure representation of [`JSONSchemaProps`](https://github.com/kubernetes/kubernetes/blob/d0c3b708026f681f7467d36dcf9696f89ccfcfab/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1/types_jsonschema.go#L20), while adding vendor extension fields for:
+* `x-kubernetes-preserve-unknown-fields: true` – recursive version of `additionalProperties: true`, influencing pruning to not prune unknown fields (compare [pruning KEP](https://github.com/kubernetes/enhancements/pull/709) for more details).
+* `x-kubernetes-embedded-resource: true` – marks a RawExtension field to embed a complete `runtime.Object` like object with `TypeMeta` and `ObjectMeta`.
+* `x-kubernetes-unions: [...]` as defined in [KEP: Union Types](https://github.com/kubernetes/enhancements/pull/926)
+* `x-kubernetes-int-or-string: true` – marks a `IntOrString` polymorphic type.
+
+Observe that any validated `JSONSchemaProps` can be represented as `StructuralSchema`. Validation allows us to drop some unsupported fields (which are part of `JSONSchemaProps` but which we forbid in CRDs):
+
+```go
+type StructuralSchema struct {
+	Items                *StructuralSchema
+	Properties           map[string]StructuralSchema
+
+	Generic
+	Extensions
+	
+	*ValueValidation
+}
+
+type Extensions struct {
+	// x-kubernetes-unions lists unions, i.e. sets of fields which are 
+	// mutual exclusive.
+	XUnions []Union
+	
+        // x-kubernetes-preserve-unknown-fields stops the API server 
+        // decoding step from pruning fields which are not specified 
+        // in the validation schema. This affects fields recursively,
```

Comment:
am I reading too much into this comment, or are you distinguishing between StructuralSchema and ValidationSchema here?

Reply:
I don't. I have removed the word "validation". Less confusion that way.